src/backend/distributed/master/worker_node_manager.c

```diff
@@ -165,6 +168,113 @@ WorkerGetRoundRobinCandidateNode(List *workerNodeList, uint64 shardId,
 
 
 /*
+ * WorkerGetLocalFirstCandidateNode takes in a list of worker nodes, and then
+ * allocates a new worker node. The allocation is performed according to the
+ * following policy: if the list is empty, the node the caller is connecting
+ * from is allocated; if the list is not empty, a node is allocated according
+ * to random policy.
+ */
+WorkerNode *
+WorkerGetLocalFirstCandidateNode(List *currentNodeList)
+{
+	WorkerNode *candidateNode = NULL;
+	uint32 currentNodeCount = list_length(currentNodeList);
+
+	/* choose first candidate node to be the client's host */
+	if (currentNodeCount == 0)
+	{
+		char *errorMessage = NULL;
+		char *remoteHostname = HostnameClientAddress(&errorMessage);
+
+		if (remoteHostname == NULL)
+		{
+			ereport(ERROR, (errmsg("%s", errorMessage),
+							errdetail("Could not find the first worker "
+									  "node for local-node-first policy."),
+							errhint("Make sure that you are not on the "
+									"master node.")));
+		}
+
+		/* if hostname is localhost.localdomain, change it to localhost */
+		if (strncmp(remoteHostname, "localhost.localdomain", WORKER_LENGTH) == 0)
+		{
+			remoteHostname = pstrdup("localhost");
+		}
+
+		candidateNode = WorkerGetNodeWithName(remoteHostname);
+		if (candidateNode == NULL)
+		{
+			ereport(ERROR, (errmsg("could not find worker node for "
+								   "hostname: %s", remoteHostname)));
+		}
+	}
+	else
+	{
+		/* find a candidate node different from those already selected */
+		candidateNode = WorkerGetRandomCandidateNode(currentNodeList);
+	}
+
+	return candidateNode;
+}
+
+
+/*
+ * HostnameClientAddress allocates memory for the connecting client's fully
+ * qualified hostname, and returns this name. If there is no such connection or
+ * the connection is over Unix domain socket, the function returns null and
+ * fills the error message.
+ */
+static char *
+HostnameClientAddress(char **errorMessage)
```

Comment:
This signature feels a little weird. How about returning an error code and appending the hostname to a StringInfo that gets passed as a parameter.